### PR TITLE
blockchain/vm: bn256 gas cost change (eip1108)

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
- This PR contains istanbul hardfork item. It is not activated yet since istanbulCompatible block number is set nil.
- This PR is derived from https://github.com/ethereum/go-ethereum/commit/beff5fa578b5349004d146d6e399aa0cca1550e5, and `core/vm/contracts.go` and `params/protocol_params.go` are referenced.
- Unlike ethereum, klaytn has `computationCost` concept, so bn256 computationCost need to be revised too. However, eip1108's purpose is to reflect the performance improvement by adopting cloudflare library, but klaytn already have this library before `computationCost` concept is introduced. So, computationCost is not revised at here.
- Main change
  - Bn256AddGas: 500 -> 150
  - Bn256ScalarMulGas: 40000 -> 6000
  - Bn256PairingBaseGas: 100000 -> 45000, Bn256PairingPerPointGas: 80000 -> 34000

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by aidan-kwon — DISMISSED** [review]
LGTM

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — DISMISSED** [review]
LGTM .

I think you can make some unit test for checking used gas. @yoomee1313

**Comment by yoomee1313** [comment]
@KimKyungup `TestBn256GasCost` test function is added at https://github.com/klaytn/klaytn/pull/987/commits/1a921ad2e0d24c25ce97a0a966fae76f69da9004 for checking bn256 used gas cost! Also, `evm_test.go` is refactored a bit due to duplication.

File: blockchain/vm/evm_test.go [hunk 1]
```diff
@@ -97,3 +56,79 @@ func TestPrecompiledContractAddressMapping(t *testing.T) {
...
+	runPrecompiledContractTestWithHFCondition(t, config, []TestData{
+		// Test whether appropriate gas cost is returned
+		// Condition 1. Caller Contract Deploy - before IstanbulCompatible, Call - before istanbulCompatible
+		{"0x006", bn256AddInput, false, Block4, params.Bn256AddGasConstantinople, bn256AddOutput, nil},
+		{"0x007", bn256ScalarMulInput, false, Block4, params.Bn256ScalarMulGasConstantinople, bn256ScalarMulOutput, nil},
+		{"0x008", bn256PairingInput, false, Block4, 260000, bn256PairingOutput, nil},
+		// Condition 2. Caller Contract Deploy - before IstanbulCompatible, Call - after istanbulCompatible
+		{"0x006", bn256AddInput, false, Block5, params.Bn256AddGasConstantinople, bn256AddOutput, nil},
+		{"0x007", bn256ScalarMulInput, false, Block5, params.Bn256ScalarMulGasConstantinople, bn256ScalarMulOutput, nil},
+		{"0x008", bn256PairingInput, false, Block5, 260000, bn256PairingOutput, nil},
```
Comment (aidan-kwon) [inline blockchain/vm/evm_test.go:128]:
```suggestion
		{"0x008", bn256PairingInput, false, Block5, Bn256PairingBaseGasConstantinople+Bn256PairingPerPointGasConstantinople*2, bn256PairingOutput, nil},
```

File: blockchain/vm/evm_test.go [hunk 2]
```diff
@@ -97,3 +56,79 @@ func TestPrecompiledContractAddressMapping(t *testing.T) {
...
+		// Condition 3. Caller Contract Deploy - after IstanbulCompatible, Call - after istanbulCompatible
+		{"0x006", bn256AddInput, true, Block5, params.Bn256AddGasIstanbul, bn256AddOutput, nil},
+		{"0x007", bn256ScalarMulInput, true, Block5, params.Bn256ScalarMulGasIstanbul, bn256ScalarMulOutput, nil},
+		{"0x008", bn256PairingInput, true, Block5, 113000, bn256PairingOutput, nil},
```
[the remark this answers is not in the document]
Comment (yoomee1313) [reply blockchain/vm/evm_test.go]: reflected at https://github.com/klaytn/klaytn/pull/987/commits/2dd3aac261b5e453d8c15589dfcc1135007acc5c

**Review by aidan-kwon — APPROVED** [review]

**Comment by yoomee1313** [comment]
@KimKyungup @ehnuje please take a look when you have a time

File: blockchain/vm/evm_test.go [hunk 3]
```diff
@@ -97,3 +56,79 @@ func TestPrecompiledContractAddressMapping(t *testing.T) {
...
+		// Condition 1. Caller Contract Deploy - before IstanbulCompatible, Call - before istanbulCompatible
+		{"0x3fd", vmLogInput, false, Block4, 0, "", kerrors.ErrPrecompiledContractAddress},
+		{"0x3fe", feePayerInput, false, Block4, 0, "", kerrors.ErrPrecompiledContractAddress},
+		{"0x3ff", validateInput, false, Block4, 0, "", kerrors.ErrPrecompiledContractAddress},
+		// Condition 2. Caller Contract Deploy - before IstanbulCompatible, Call - after istanbulCompatible
+		{"0x3fd", vmLogInput, false, Block5, 0, "", kerrors.ErrPrecompiledContractAddress},
+		{"0x3fe", feePayerInput, false, Block5, 0, "", kerrors.ErrPrecompiledContractAddress},
+		{"0x3ff", validateInput, false, Block5, 0, "", kerrors.ErrPrecompiledContractAddress},
+		// Condition 3. Caller Contract Deploy - after IstanbulCompatible, Call - after istanbulCompatible
+		{"0x009", vmLogInput, true, Block5, math.MaxUint64, "", errors.New("invalid input length")},
```
Comment (KimKyungup) [inline blockchain/vm/evm_test.go:99]: @yoomee1313 Why does this test case have `math.MaxUint64` gas?

**Review by KimKyungup — APPROVED** [review]

File: blockchain/vm/evm_test.go (hunk 3 above)
Comment (yoomee1313) [reply blockchain/vm/evm_test.go:99]: @KimKyungup `invalid input length` error occurs when executing blake2f precompiled contract. That is, it is a VM error, and if any vm error occurs, all remaining gas is consumed. In this tc, gas limit is Uint64, so it consumes MaxUint64.

> // When an error was returned by the EVM or when setting the creation code
> // above we revert to the snapshot and consume any gas remaining. Additionally
> // when we're in homestead this also counts for code storage gas errors.

**Comment by aidan-kwon** [comment]
@yoomee1313 Let's merge this PR.

**Review by ehnuje — APPROVED** [review]

**Pull request merged**